src/importer.cpp

```diff
@@ -0,0 +1,483 @@
+        if (!importSource->hasModel()) {
+            std::string url = importSource->url();
+            if (mLibrary.count(url) == 0) {
+                url = resolvePath(importSource->url(), baseFile);
+            }
+            if (mLibrary.count(url) == 0) {
+                // If the url has not been resolved into a model in this library, parse it and save.
+                std::ifstream file(url);
+                if (file.good()) {
+                    std::stringstream buffer;
+                    buffer << file.rdbuf();
+                    auto parser = Parser::create();
+                    auto model = parser->parseModel(buffer.str());
+                    importSource->setModel(model);
+
+                    // Save the pair of model and url to the library map.
+                    mLibrary.insert(std::make_pair(url, model));
+
+                    // Save the imported model to a list of external dependencies.
+                    mExternals.emplace_back(std::make_pair(url, importedEntity->importReference()));
+                    return doResolveImports(model, url, history);
+                }
+            } else {
+                // If it has, then pass the previous model instance to the import source.
+                auto model = mLibrary[url];
+                importSource->setModel(model);
+                if (typeString == "component") {
+                    return checkComponentForCycles(model, history);
+                }
+                return checkUnitsForCycles(model, history);
+            }
+        }
+    }
+    return true;
+}
+
+bool Importer::ImporterImpl::resolveComponentImports(const ComponentEntityPtr &parentComponentEntity,
+                                                     const std::string &baseFile,
+                                                     std::vector<std::tuple<std::string, std::string, std::string>> &history)
+{
+    bool noErrors = true;
+    for (size_t n = 0; n < parentComponentEntity->componentCount(); ++n) {
+        libcellml::ComponentPtr component = parentComponentEntity->component(n);
+        if (component->isImport()) {
+            if (!resolveImport(component, component->name(), "component", baseFile, history)) {
+                if (!history.empty()) {
+                    std::string msg = "Cyclic dependencies were found when attempting to resolve components";
+                    auto parentModel = owningModel(component);
+                    if (parentModel != nullptr) {
+                        msg += " in model '" + parentModel->name() + "'";
+                    }
+                    msg += ". The dependency loop is:\n";
+                    std::string spacer = "    ";
+                    for (auto &h : history) {
+                        msg += spacer + "component '" + std::get<0>(h) + "' imports '" + std::get<1>(h) + "' from '" + std::get<2>(h);
+                        spacer = "',\n    ";
+                    }
+                    msg += "'.";
```

Comment:
Is there really no way to refactor this code so that it doesn't overlap with https://github.com/cellml/libcellml/pull/608/files#r464766862? In any case, I would reformat the warning message.

Reply:
Perhaps, but this PR is about moving code from one place to another as a first go.  Other improvements can come later.

Reply:
Ths is new code that can easily be (and should be) refactored. Worst case scenario, please create an issue for it. It will just be one more issue...